### PR TITLE
Using the simplest linters from lintr to check for style.

### PR DESCRIPTION
**Pull request by mb706**
Re #1599
Some problems become apparent:
1. tests are also tested, byt don't adhere to style guidelines.
2. exponentiation is an infix-operator that shouldn't need spaces.
3. noncomliant library functions that we can't do anything give errors.
4. df slots with underscores exist

### CODE REVIEW

**Comment by berndbischl** [comment]
ok we should review this with a few persons for some minutes

**Comment by mb706** [comment]
Worrying that travis tests pass, isn't it?

**Comment by jakob-r** [comment]
I think all occurrences of `some.function(foo <- bar)` can be replaced by `some.function({foo = bar})`. Should that be part of this PR?

**Comment by larskotthoff** [comment]
Yes, I would make it part of this PR.

**Comment by mb706** [comment]
If you edit code to adhere more to the style guide, you should do it in the master branch (or in its own PR that gets merged quickly). Then, whenever the style adherence in master improved, master can be merged back into the lintr PR (the one you're looking at right now). lintr can then be merged into master when all the style guides pass.

It would not be a good idea to collect *all* style corrections in here, because that would take a long time, and over time, it would accumulate merge conflicts.

Please, everyone feel free to improve the lintr functionality. Small TODO list:

- [ ] `line_length_linter`
- [ ] somehow check for two spaces indentation per parenthesis / braces level
- [ ] check that `T` and `F` don't get used
- [ ] check `c(1:3)` is not used. (I guess check that `c(` is always followed by at least one comma before `)`)
- [ ] improve `infix_spaces_linter` to accept exponentiation with no spaces. (I assume `2^x` is legal, even if not explicitly stated in the style guide?)
- [ ] test for and forbid `== TRUE`
- [ ] test for `;`
- [ ] check indentation of comments, check that `#` is followed by a space or a single quote
- [ ] check that functions are camel case and that non-functions are not. There is a `camel_case_linter`, maybe it can be adapted.
- [ ] `snake_case_linter` seems to be broken right now, since it complains about imported functions sometimes. There is some functionality in lintr to work against this, but it seems imperfect. Maybe there is a way to only check for variable name guides if the variable appears to the lest of an `EQ_ASSIGN`.
- [ ] check for absence of TODO.

Things that are not in the style guide but would make sense to me:
- [ ] `spaces_insice_linter`
- [ ] `trailing_blank_lines_linter`
- [ ] `trailing_whitespace_linter`
- [ ] `seq_linter`
- [ ] test for `v*apply` instead of `sapply` (h/t @larskotthoff )
- [ ] check no more than two consecutive empty lines?

**Comment by mb706** [comment]
`Error in library("lintr") : there is no package called 'lintr'`
Anyone knows why this happens?

**Comment by berndbischl** [comment]
> Error in library("lintr") : there is no package called 'lintr'
> Anyone knows why this happens?


is it really standard to combine lintr with testthat in the way you did (just asking)?
all packages you use in testthat tests, must be in SUGGESTS.

**Comment by mb706** [comment]
lintr's doc uses
```R
if (requireNamespace("lintr", quietly = TRUE)) {
  ...
}
```
but I want the test to fail if lintr for some reason is not available.

**Comment by mb706** [comment]
I guess it works now

File: tests/testthat/test_lint.R [hunk 1]
```diff
@@ -0,0 +1,73 @@
+library("lintr")
+context("lint")
+
+# prohibit <-
+left_assign_linter = function(source_file) {
+  lapply(lintr:::ids_with_token(source_file, "LEFT_ASSIGN"), function(id) {
+      parsed <- lintr:::with_id(source_file, id)
```
Comment (larskotthoff) [inline tests/testthat/test_lint.R]: Ahhhh, the style guide checker doesn't follow the style guide!

Comment (ja-thomas) [inline tests/testthat/test_lint.R]: lol

Comment (Stevo15025) [inline tests/testthat/test_lint.R]: who watches the watchers

**Comment by mb706** [comment]
Is Jenkins build failing because it uses an old lintr version?

**Comment by larskotthoff** [comment]
No, current version is installed. The last travis build to complete failed as well.

**Comment by mb706** [comment]
Jenkins appears to use the CRAN version of lintr, while Travis uses the github version. The CRAN version apparently hast he problem that it complains about `[, , 1]`, because there is supposed to be *no space before a comma*, but also about `[,, 1]`, because there is always supposed to be a *space after a comma*. Is there a way to make Jenkins use the github version as well, or is this generally discouraged?

(Travis seems to fail for some dumb, but different, reason as well. Apparently I cannot view the job logs of the Travis builds?)

**Comment by jakob-r** [comment]
You would have to adapt the jenkins script. But should we bother, that it fails on Jenkins?

**Comment by mb706** [comment]
### Introducing `quicklint`
The complete linting takes quite a while. I wrote a script that asks `git` which files have changed compared to the `master` branch differ and lints only these. The R check still runs the complete lint, but this is a quick check before uploading things to github and waiting 45 min for Travis.

Just run the `todo-files/quicklint` script. (Does this file belong somewhere else?) An optional argument is the top path of `mlr`; by default, `quicklint` uses the path that `git` also uses (i.e. the first directory that contains a `.git` directory), so one can run `quicklint` anywhere within the mlr directory structure.

This works on my Linux machine, maybe some Mac person wants to check if it also works for them.

**Comment by mb706** [comment]
I cannot look at travis build logs. Maybe its an adblocker thing; does [https://travis-ci.org/mlr-org/mlr/builds/214248340](https://travis-ci.org/mlr-org/mlr/builds/214248340) work for someone?

**Comment by mb706** [comment]
As soon as I get the Travis build passing I am done here.
1. Can we merge this? The PR touches a lot of files (expectedly), so I would expect lots of merge conflicts if many things are changed before this is merged.
2. I don't know how to get Jenkins to use the github version of Travis; if someone else could to that it would be nice.

**Comment by larskotthoff** [comment]
I've installed the github version of lintr on Jenkins. For scripts, I use the `thirdparty` directory.

**Comment by mb706** [comment]
Thanks!

**Comment by mb706** [comment]
I'm done with this, imho this can be merged.

**Comment by larskotthoff** [comment]
Great, thanks! I'll merge this now so we don't get conflicts.

**Pull request merged**